pyo3-macros-backend/src/pymethod.rs

```diff
@@ -987,17 +989,23 @@ fn extract_object(
 enum ReturnMode {
     ReturnSelf,
     Conversion(TokenGenerator),
+    SpecializedConversion(TokenGenerator, TokenGenerator),
 }
 
 impl ReturnMode {
     fn return_call_output(&self, call: TokenStream) -> TokenStream {
         match self {
             ReturnMode::Conversion(conversion) => quote! {
-                let _result: _pyo3::PyResult<#conversion> = #call;
+                let _result: _pyo3::PyResult<#conversion> = _pyo3::callback::convert(py, #call);
                 _pyo3::callback::convert(py, _result)
             },
+            ReturnMode::SpecializedConversion(traits, tag) => quote! {
+                let _result = #call;
+                use _pyo3::callback::{#traits};
+                (&_result).#tag().convert(py, _result)
+            },
```

Comment:
@davidhewitt This is by no means complete but I wanted to run the design by you before I go off into the weeds.

So as per the discussion in #3190, I tried to remove any special handling from `__next__` so that if user code wants to stop iteration, I can just return a `PyStopIteration` error. `Option` is still handled separately though using autoref-based specialization. Here, I opted to deprecate the format but I am not actually sure whether we want to fully deprecate this or just keep the specialized behaviour?

Furthermore, I am not sure whether I messed up the return value conversion by delaying the call to `convert` but this was the only way I was able to make type inference work out. I also wonder whether `PyResult<Option<T>>` is another form we should consider supporting now?

Reply:
I think delaying the call to `convert` is fine in all cases except when we use temporary `PyRef` and `PyRefMut` in the expression to extract `&T` and `&mut T` class types.

I wonder whether we could abuse `FnOnce::Output` to inspect the type for tag inference before we actually create the value. I can try to express that further when I'm back at my computer.

Reply:
> I think delaying the call to convert is fine in all cases except when we use temporary PyRef and PyRefMut in the expression to extract &T and &mut T class types.

So this is with regards to extending the lifetime of temporaries? So the test test case would be a `__next__` implementation taking `&T` where `T: PyClass` as a parameter?

> I wonder whether we could abuse FnOnce::Output to inspect the type for tag inference before we actually create the value. I can try to express that further when I'm back at my computer.

I am not sure as autoref-based specialization cannot work on types alone, it needs references to values. We could probably try to create these references out of thin air but I fear we would run into safety/validity issues, i.e. insta-UB.

Reply:
> So the test test case would be a __next__ implementation taking &T where T: PyClass as a parameter?

I think it's needs to take `&T` and also return data borrowed from that `T` back out of the function. For example, try replacing `__next__` in `test_proto_methods.rs` line 661 with this implementation and you'll get a compile error which #3666 fixes:

```rust
    fn __next__<'py>(&'py mut self, py: Python<'py>) -> Option<&'py PyAny> {
        if !self.polled {
            self.polled = true;
            Some(self.future.as_ref(py))
        } else {
            None
        }
    }
```

Reply:
Modified the test case and with the holders from #3667 it now compiles.